ts/parsers/paragraphparser.ts

```diff
@@ -1,37 +1,155 @@
-                };
-            });
+        const paras = [];
+        const stack = [];
+        const paragraph: Paragraph = {
+            list: {
+                listType: ListType.Ordered,
+                listItems: []
+            }
+        };
+        let currentParagraph = paragraph;
+        let currentLevel = -1;
 
-            return {
-                content: content,
-                paragraphProperties: this.determineParagraphProperties(paragraph)
-            };
-        });
+        for (const p of paragraphs) {
+            if (this.isList(p)) {
+                const listLevel = this.getListlevel(p);
+
+                //if its the first of the list kind
+                if (currentLevel === -1) {
+                    currentLevel = 0;
+                    currentParagraph.list.listType = this.getListType(p);
+                    currentParagraph.list.listItems.push(this.getParagraph(p));
+                    stack.push(currentParagraph);
+                }
+                //if the level is same keep pushing the list items in the same array
+                else if (listLevel === currentLevel) {
+                    currentParagraph.list.listItems.push(this.getParagraph(p));
+                } else if (listLevel > currentLevel) {
+                    //edge case to handle if multiple levels are jumped ahead
+                    //create empty paragraphs/lists to maintain hierarchy and fill in the level gaps
+                    while (currentLevel < listLevel - 1) {
+                        const emptyPara: Paragraph = {
+                            list: {
+                                listType: ListType.UnOrdered,
+                                listItems: []
+                            }
+                        };
+                        currentParagraph.list.listItems.push(emptyPara);
+                        currentParagraph = emptyPara;
+                        //pushing it in the stack to keep track of the parents
+                        stack.push(emptyPara);
+                        currentLevel++;
+                    }
+                    //if there is another hierarchy starting create a new list for it
+                    const newPara: Paragraph = {
+                        list: {
+                            listType: this.getListType(p),
+                            listItems: [this.getParagraph(p)]
+                        }
+                    };
+                    currentParagraph.list.listItems.push(newPara);
+                    currentParagraph = newPara;
+                    //pushing it in the stack to keep track of the parents
+                    stack.push(newPara);
+                    currentLevel++;
+                } else {
+                    //if we find the list level lower than current level
+                    //keep going back in stack until the same level parent found
+                    while (currentLevel !== listLevel) {
```

Comment:
you can use < comparison inside then while and it will self document what it is doing .. also the `else` above can be replaced with an if with the exact check on `if (listLevel < currentLevel)`

Reply:
yea that can be changed to `<`. However not really a fan of just using ifs, even if we know that its not going to be called. 
as else will give a clear visibility that this part will not be executed, which avoids confusions.
Because we only need to call exactly one of these in one iteration. there is no use case and it shouldn't run multiple ifs in one iteration. 
Like previously mentioned in one of the previous comments, it did create a problem for me. having duplicates as another if was called before.

Reply:
>Because we only need to call exactly one of these in one iteration
then you can exit early and then doesn't matter if you have more ifs later on, they will not be called.

incrementing/decrementing global variables will be a problem and I can see that, returning early (`continue;`) will solve that as well tbh

Reply:
changed the while condition, to be `currentLevel > listLevel` This made more sense to me. as we decrement the currentLevel. 
...Wese baat aik hi hay kaan idhr se pakro ya udhr se 😄

Reply:
lol haan .. just being pedantic but it does improve readability, it all adds up. Thanks!